### PR TITLE
Add DialContext function

### LINKED ISSUE #475: Add DialContext function
I want Context to be used when dialing.
So I want DialContext.

By default you should use Dialer.DialContext.
https://golang.org/pkg/net/#Dialer.DialContext

### PR DESCRIPTION
**Pull request by 178inaba**
Fixes #475.

### CODE REVIEW

File: redis/conn.go [hunk 1]
```diff
@@ -127,6 +142,15 @@ func DialNetDial(dial func(network, addr string) (net.Conn, error)) DialOption {
 	}}
 }
 
+// DialNetDialContext specifies a custom dial function with context for creating TCP
+// connections, otherwise a net.Dialer customized via the other options is used.
+// DialNetDialContext overrides DialConnectTimeout and DialKeepAlive.
+func DialNetDialContext(dialContext func(ctx context.Context, network, addr string) (net.Conn, error)) DialOption {
```
Comment (stevenh) [inline redis/conn.go:148]: A more idiomatic name for this would be DialWithContext

File: redis/conn.go [hunk 2]
```diff
@@ -78,6 +79,7 @@ type dialOptions struct {
 	writeTimeout time.Duration
 	dialer       *net.Dialer
 	dial         func(network, addr string) (net.Conn, error)
```
Comment (stevenh) [inline redis/conn.go:81]: As this is internal we could always use dialContext with a context.Background if need.

File: redis/conn.go [hunk 3]
```diff
@@ -127,6 +142,15 @@ func DialNetDial(dial func(network, addr string) (net.Conn, error)) DialOption {
 	}}
 }
 
+// DialNetDialContext specifies a custom dial function with context for creating TCP
+// connections, otherwise a net.Dialer customized via the other options is used.
+// DialNetDialContext overrides DialConnectTimeout and DialKeepAlive.
+func DialNetDialContext(dialContext func(ctx context.Context, network, addr string) (net.Conn, error)) DialOption {
+	return DialOption{func(do *dialOptions) {
+		do.dialContext = dialContext
```
Comment (stevenh) [inline redis/conn.go:150]: We could simplify things by only using a dialer with context, as mentioned above.

Comment (178inaba) [reply redis/conn.go:138]: Fix!

File: redis/conn.go (hunk 2 above)
Comment (178inaba) [reply redis/conn.go]: Fix!

File: redis/conn.go (hunk 1 above)
Comment (178inaba) [reply redis/conn.go:136]: Do I have to be consistent with DialNetDial?
https://pkg.go.dev/github.com/gomodule/redigo/redis?tab=doc#DialNetDial

Dialer uses DialContext.
https://golang.org/pkg/net/#Dialer.DialContext

There is an example to use WithContext when it takes the context itself as an argument.
https://golang.org/pkg/net/http/#Request.WithContext

Comment (stevenh) [reply redis/conn.go:136]: You are right that `DialNetDialContext` follows the precedent of `DialNetDial` but I think it's quite a mouth full.

What do you think users would prefer?

Comment (178inaba) [reply redis/conn.go]: What about `DialContextFunc`?

**Comment by 178inaba** [comment]
@stevenh I added a test!

**Comment by 178inaba** [comment]
@stevenh Hello!
Can you review this pull request again?

**Pull request merged**